### ISSUE TITLE
OnGetChildProperty and OnSetChildProperty don't seem to be triggered when I get/set a child property of a container

### ISSUE THREAD

**Issue by NepNet** [GtkSharp/GtkSharp#266]
I was trying out something and I wanted to make a thing to sort objects inside a box when a button is pressed and wanted to make my test application to be doing something when the child property position for GtkBox is updated but the event doesn't trigger I wanted to hook into the OnChildSetProperty instead of the Clicked event of the button because I want to trigger the said procedure when the children of the box are moved no mater what the cause may be ( a button, GtkInspector, progamatically setting the position property) the functions to set/get properties seem to work but the callbacks don't. Also I wanted to add custom child properties which is also missing, I have a local branch with that function that is needed for this but it requires those 2 callbacks to work and looking at the source code I couldn't figure out how to get them to fire when childget/setproperty are called

Sorry if my post is hard to read I tried to be as concise as I could

**Comment by lytico** [comment]
maybe a link to your project would be helpful? or  code example ?

**Comment by zii-dmg** [comment]
I tried GtkSharp and native C test app, doesn't work. I don't understand why. Source code of GtkBox (and other Gtk widgets) using it so it should work.

**Comment by harry-cpp** [comment]
Even with `Glib.ConnectBefore`?

**Comment by zii-dmg** [comment]
Its virtual function, not signal. GtkSharp (and C) just replace function pointer. So I guest ConnectBefore is not relevant.

**Comment by NepNet** [comment]
After some investigation extending the Box class and overriding these virtual methods won't work for catching child properties that aren't owned by the class(the design was flawed from the start anyway). In the meantime I was working on implementing the InstallChildProperty function that seems to be missing from the bindings and as @zii-dmg said I have to replace the function pointer for these 2 "signals" which is done via a private function in the container class also the delegates for these aren't marked public either also that function to set the callbacks for this isn't called anywhere as my IDE says. So overriding these 2 virtual methods wouldn't do anything because the callbacks aren't set for them.

TLDR: I made a fork and put my weird hacky way to let me override the GetChildProperty and SetChildProperty callbacks which can induce bugs if used on classes that already implement them.

**Comment by zii-dmg** [comment]
Generated virtual functions is replaced by GtkSharp automatically.
```csharp
[GLib.DefaultSignalHandler(Type=typeof(Gtk.Container), ConnectionMethod="OverrideSetChildProperty")]
protected virtual void OnSetChildProperty (Gtk.Widget child, uint property_id, GLib.Value value, IntPtr pspec)
{
    InternalSetChildProperty (child, property_id, value, pspec);
}

static void OverrideSetChildProperty (GLib.GType gtype, SetChildPropertyNativeDelegate callback)
{
    unsafe {
        IntPtr* raw_ptr = (IntPtr*)(((long) gtype.GetClassPtr()) + (long) class_abi.GetFieldOffset("set_child_property"));
        *raw_ptr = Marshal.GetFunctionPointerForDelegate((Delegate) callback);
    }
}
```

`OverrideSetChildProperty` is called on first access to your type. See it gets ptr to C struct with pointers, offsets to where needed fuction pointer is located and writes to it our delegate pointer.

**Comment by lytico** [comment]
Container.GetChildProperty is glue-code in gtk-sharp: 
https://github.com/mono/gtk-sharp/blob/dadc19cf1b90c5743f2776c675faac990e397a56/gtk/glue/container.c#L66

**Comment by NepNet** [comment]
Seems like the problem of why the OnGetChildProperty and OnSetChildProperty not being called was because if I inherit from a class these properties don't get associated with the class based on it. They seem to be triggered if I add a child property to the class problem is that there seems to be no way to add child properties. I have on my fork that function to install child properties and gtk doesn't throw ` gtk_container_class_install_child_property: assertion 'cclass->set_child_property != NULL' failed` if I override the virtual method for that(OnSetChildProperty) and by using gtk inspector to debug I can confirm it works however for OnGetChildProperty it doesn't seem to change the displayed value in the inspector
This is the code I use 
`protected override void OnGetChildProperty(Widget child, uint property_id, Value value, IntPtr pspec)
{
    value = new Value(20);
    base.OnGetChildProperty(child, property_id, value, pspec);
}`
and yes for the OnGetChildProperty callback you have to send the value to the callback because it is called when the value of a property is requested
